File: app/src/main/java/com/anadi/weatherinfo/repository/LocationsCash.java
package com.anadi.weatherinfo.repository;

import android.content.Context;

import androidx.annotation.Nullable;

import com.anadi.weatherinfo.addlocation.AddLocationContract;
import com.anadi.weatherinfo.addlocation.LocationsProvider;
import com.anadi.weatherinfo.details.DetailsContract;
import com.anadi.weatherinfo.mainactivity.MainActivityContract;
import com.anadi.weatherinfo.repository.data.WeatherInfo;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.Observable;

import timber.log.Timber;

public class LocationsCash extends Observable implements AddLocationContract.Model, MainActivityContract.Model,
        DetailsContract.Model {

    private static final String APP_DATA_FILE = "weatherinfo.db";
    private static final String APP_DATA_DIR = "db";
    private static final int TIME_TO_UPDATE = 1000 * 60 * 60; // data is considered to be "fresh" during one hour
    // Singleton with double check
    private static volatile LocationsCash instance;
    private final LocationsProvider locationsProvider = Locations.getInstance();
    private final InfoLoader infoLoader = InfoLoader.getInstance();
    private ArrayList<LocationInfo> locations = new ArrayList<>();

    private LocationsCash() {}

    public static LocationsCash getInstance() {
        LocationsCash result = instance;
        if (result == null) {
            synchronized (LocationsCash.class) {
                result = instance;
                if (result == null) {
                    instance = result = new LocationsCash();
                }
            }
        }
        return result;
    }

    @Override
    public boolean add(String cityName, String countryName) {
        Country country = locationsProvider.getCountryByName(countryName);

        if (country == null) {
            Timber.d("There is no such country: %s", countryName);
            return false;
        }

        LocationInfo locationInfo = getCashedInfo(cityName, country);
        if (locationInfo != null) {
            Timber.d("City already loaded: %s", cityName);
            return true;
        }

        return load(cityName, country);
    }

    @Override
    public void loadLocations(Context context) {
        locationsProvider.loadLocations(context);
    }

    @Override
    public void saveData(Context context) {
        File path = new File(context.getFilesDir(), APP_DATA_DIR);
        if (!path.exists()) {
            if (!path.mkdirs()) {
                Timber.d("Failed to create app db dir!");
                return;
            }
        }

        File file = new File(path, APP_DATA_FILE);
        if (!file.exists()) {
            try {
                if (!file.createNewFile()) {
                    Timber.d("Failed to create app db file!");
                    return;
                }
            } catch (IOException e) {
                Timber.d("Failed to create app db file!");
                e.printStackTrace();
                return;
            }
        }

        try (FileOutputStream fos = new FileOutputStream(file); ObjectOutputStream oos = new ObjectOutputStream(fos)) {

            oos.writeObject(locations);
            oos.flush();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void loadData(Context context) {
        File path = new File(context.getFilesDir(), APP_DATA_DIR);
        if (!path.exists()) {
            Timber.d("No db dir!");
            return;
        }

        File file = new File(path, APP_DATA_FILE);
        if (!file.exists()) {
            Timber.d("No db file!");
            return;
        }

        try (FileInputStream fis = new FileInputStream(file); ObjectInputStream ois = new ObjectInputStream(fis)) {

            locations = (ArrayList<LocationInfo>) ois.readObject();

        } catch (ClassNotFoundException | IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public ArrayList<LocationInfo> getLocations() {
        return locations;
    }

    @Override
    public boolean deleteLocation(LocationInfo locationInfo) {
        boolean result = locations.remove(locationInfo);
        if (result) {
            setChanged();
            notifyObservers();
        }

        return result;
    }

    @Override
    public @Nullable
    WeatherInfo getInfo(int id) {

        for (LocationInfo locationInfo : locations) {
            if (locationInfo.getId() == id) {
                return locationInfo.getInfo();
            }
        }

        return null;
    }

    @Override
    public boolean update(int id) {
        for (LocationInfo locationInfo : locations) {
            if (locationInfo.getId() == id) {

                WeatherInfo weatherInfo = infoLoader.load(locationInfo.getCityName(), locationInfo.getCountry());
                if (weatherInfo == null) {
                    Timber.d("No info loaded for location: " + locationInfo);
                    return false;
                }

                locationInfo.setInfo(weatherInfo);
                setChanged();
                notifyObservers(locationInfo);
                return true;
            }
        }

        return false;
    }

    @Override
    public boolean needUpdate(int id) {
        for (LocationInfo locationInfo : locations) {
            if (locationInfo.getId() == id) {
                return dataIsOld(locationInfo.getInfo().dt);
            }
        }

        Timber.d("There is no data at all for id = %d", id);
        return false;
    }

    private boolean load(final String cityName, final Country country) {

        WeatherInfo weatherInfo = infoLoader.load(cityName, country);

        if (weatherInfo == null) {
            Timber.d("No info loaded for location: %s, %s", cityName, country);
            return false;
        }

        final LocationInfo locationInfo = new LocationInfo(cityName, country);
        locationInfo.setInfo(weatherInfo);
        locations.add(locationInfo);
        setChanged();
        notifyObservers();

        return true;
    }

    private @Nullable
    LocationInfo getCashedInfo(final String cityName, final Country country) {

        for (LocationInfo locationInfo : locations) {
            if (locationInfo.getCityName().equals(cityName) && locationInfo.getCountry().equals(country)) {
                return locationInfo;
            }
        }

        return null;
    }

    private boolean dataIsOld(int timestamp) {
        return System.currentTimeMillis() - timestamp > TIME_TO_UPDATE;
    }
}

File: app/src/main/java/com/anadi/weatherinfo/repository/LocationsCash.kt
package com.anadi.weatherinfo.repository

import android.content.Context
import com.anadi.weatherinfo.repository.data.WeatherInfo
import com.anadi.weatherinfo.ui.addlocation.AddLocationContract
import com.anadi.weatherinfo.ui.addlocation.LocationsProvider
import com.anadi.weatherinfo.ui.details.DetailsContract
import com.anadi.weatherinfo.ui.mainactivity.MainActivityContract
import com.anadi.weatherinfo.utils.WeatherException
import timber.log.Timber
import java.io.*
import java.util.*
import javax.inject.Inject
import kotlin.jvm.Throws

class LocationsCash @Inject constructor (private val locationsProvider: LocationsProvider,
                                         private val infoLoader: InfoLoader) :
        Observable(),
        AddLocationContract.Model,
        MainActivityContract.Model,
        DetailsContract.Model {

    override var locations = ArrayList<LocationInfo>()
        private set

    override fun add(cityName: String, countryName: String) {
        val country = locationsProvider.getCountryByName(countryName)

        val locationInfo = getCashedInfo(cityName, country)
        if (locationInfo != null) {
            Timber.d("City already loaded: %s", cityName)
            return
        }

        load(cityName, country)
    }

    @Throws(WeatherException::class)
    private fun load(cityName: String, country: Country) {
        val weatherInfo = infoLoader.load(cityName, country)
        val locationInfo = LocationInfo(cityName, country)
        locationInfo.info = weatherInfo
        locations.add(locationInfo)
        setChanged()
        notifyObservers()
    }

    override fun loadLocations(context: Context) {
        Timber.i("loadLocations called")
        locationsProvider.loadLocations(context)
    }

    override fun saveData(context: Context) {
        val path = File(context.filesDir, APP_DATA_DIR)
        if (!path.exists()) {
            if (!path.mkdirs()) {
                Timber.d("Failed to create app db dir!")
                return
            }
        }
        val file = File(path, APP_DATA_FILE)
        if (!file.exists()) {
            try {
                if (!file.createNewFile()) {
                    Timber.d("Failed to create app db file!")
                    return
                }
            } catch (e: IOException) {
                Timber.d("Failed to create app db file!")
                Timber.e(e)
                return
            }
        }
        try {
            FileOutputStream(file).use { fos ->
                ObjectOutputStream(fos).use { oos ->
                    oos.writeObject(locations)
                    oos.flush()
                }
            }
        } catch (e: IOException) {
            Timber.e(e)
        }
    }

    override fun loadData(context: Context) {
        val path = File(context.filesDir, APP_DATA_DIR)
        if (!path.exists()) {
            Timber.d("No db dir!")
            return
        }
        val file = File(path, APP_DATA_FILE)
        if (!file.exists()) {
            Timber.d("No db file!")
            return
        }
        try {
            FileInputStream(file).use { fis -> ObjectInputStream(fis).use { ois -> locations = ois.readObject() as ArrayList<LocationInfo> } }
        } catch (e: ClassNotFoundException) {
            Timber.e(e)
        } catch (e: IOException) {
            Timber.e(e)
        }
    }

    override fun deleteLocation(locationInfo: LocationInfo): Boolean {
        val result = locations.remove(locationInfo)
        if (result) {
            setChanged()
            notifyObservers()
        }
        return result
    }

    override fun getInfo(id: Int): WeatherInfo? {
        for (locationInfo in locations) {
            if (locationInfo.id == id) {
                return locationInfo.info
            }
        }
        return null
    }

    override fun update(id: Int) {
        for (locationInfo in locations) {
            if (locationInfo.id == id) {
                tryUpdate(locationInfo)
            }
        }
    }

    @Throws(WeatherException::class)
    private fun tryUpdate(locationInfo: LocationInfo) {
        val weatherInfo = infoLoader.load(locationInfo.cityName, locationInfo.country)
        locationInfo.info = weatherInfo
        setChanged()
        notifyObservers(locationInfo)
    }

    override fun needUpdate(id: Int): Boolean {
        for (locationInfo in locations) {
            if (locationInfo.id == id) {
                return dataIsOld(locationInfo.info.dt)
            }
        }
        Timber.d("There is no data at all for id = %d", id)
        return false
    }

    private fun getCashedInfo(cityName: String?, country: Country): LocationInfo? {
        for (locationInfo in locations) {
            if (locationInfo.cityName == cityName && locationInfo.country == country) {
                return locationInfo
            }
        }
        return null
    }

    private fun dataIsOld(timestamp: Int): Boolean {
        return System.currentTimeMillis() - timestamp > TIME_TO_UPDATE
    }

    companion object {
        private const val APP_DATA_FILE = "weatherinfo.db"
        private const val APP_DATA_DIR = "db"
        private const val TIME_TO_UPDATE = 1000 * 60 * 60 // data is considered to be "fresh" during one hour
    }
}